packages/components/src/modifiers/hds-code-editor.ts

```diff
@@ -0,0 +1,189 @@
+    async (language?: HdsCodeEditorLanguages) => {
+      if (language === undefined) {
+        return;
+      }
+
+      let module: CodemirrorLanguageModule | null = null;
+      let languageFunction = null;
+
+      switch (language) {
+        case 'go':
+          module = await import('@codemirror/lang-go');
+          languageFunction = (module as CodemirrorGoModule).go;
+          break;
+        case 'json':
+          module = await import('@codemirror/lang-json');
+          languageFunction = (module as CodemirrorJsonModule).json;
+          break;
+        case 'sql':
+          module = await import('@codemirror/lang-sql');
+          languageFunction = (module as CodemirrorSqlModule).sql;
+          break;
+        case 'hcl':
+          module = await import('codemirror-lang-hcl');
+          languageFunction = (module as CodemirrorHclModule).hcl;
+          break;
+        default:
+          throw new Error(`Language ${language} is not supported`);
+      }
+
+      return languageFunction();
+    }
+  );
+
+  setupTask = task(
+    { drop: true },
+    async (
+      element: HTMLElement,
+      _positional: PositionalArgs<HdsCodeEditorSignature>,
+      named: NamedArgs<HdsCodeEditorSignature>
+    ) => {
+      const { onInput, onSetup, language, value } = named;
+
+      const [
+        {
+          EditorView,
+          keymap,
+          lineNumbers,
+          highlightActiveLineGutter,
+          highlightSpecialChars,
+          highlightActiveLine,
+        },
+        { EditorState },
+        { defaultKeymap, history, historyKeymap },
+        { bracketMatching, syntaxHighlighting },
+      ] = await Promise.all([
+        import('@codemirror/view'),
+        import('@codemirror/state'),
+        import('@codemirror/commands'),
+        import('@codemirror/language'),
+      ]);
+
+      this.onInput = onInput;
+
+      const languageExtension = await this.loadLanguageTask.perform(language);
+
+      let extensions = [
+        lineNumbers(),
+        highlightActiveLineGutter(),
+        highlightSpecialChars(),
+        highlightActiveLine(),
+        EditorView.updateListener.of((update: ViewUpdate) => {
+          // toggle a class if the update has/does not have a selection
+          if (update.selectionSet) {
+            update.view.dom.classList.toggle(
+              'cm-hasSelection',
+              !update.state.selection.main.empty
+            );
+          }
+
+          // call the onInput callback if the document has changed
+          if (!update.docChanged || this.onInput === undefined) {
+            return;
+          }
+          this.onInput(update.state.doc.toString());
+        }),
+        hdsDarkTheme,
+        keymap.of([...defaultKeymap, ...historyKeymap]),
+        bracketMatching(),
+        syntaxHighlighting(hdsDarkHighlightStyle),
+        history(),
+      ];
+
+      if (languageExtension !== undefined) {
+        extensions = [languageExtension, ...extensions];
+      }
```

Comment:
thought: this can be extracted into a separate `buildExtensions` method; not a priority, can be refactored afterwards.

Reply:
Good idea!